colossalai/utils/model/experimental.py

```diff
@@ -50,140 +58,111 @@ class LazyTensor(torch.Tensor):
-
-        Returns:
-            MetaTensor: The generated ``MetaTensor``.
-        """
-        if isinstance(self, nn.Parameter):
-            return nn.Parameter(self._meta_data, requires_grad=self.requires_grad)
-        else:
-            return self._meta_data
-
-    def distribute(self) -> torch.Tensor:
-        """Distribute the ``LazyTensor`` according to the ``ShardingSpec``.
-
-        Returns:
-            torch.Tensor: The sharded tensor.
+    def clean(self) -> None:
+        """Clean all stored operations, meta data and materialized data, which prevents memory leaking. This should be called after all tensors are materialized.
         """
-        if self._spec is None:
-            raise RuntimeError('ShardingSpec is not set for\n{self}')
-        spec, device_mesh = self._spec, self._spec.device_mesh
-        target = self.materialize()
-
-        # TODO(some man): better not be coupled with auto-parallel
-        target.data = scm.apply_for_autoparallel_runtime(target.data, ShardingSpec(device_mesh, target.shape, {}),
-                                                         spec).detach().clone()
-        return target
-
-    def _realize_cached_data(self) -> torch.Tensor:
-        # self._cached_data should be generated after the first call of this function
-        if self._cached_data is None:
-            if self._factory_method is not None:
-                # apply factory method
-                func, args, kwargs = self._factory_method
+        self._factory_method = None
+        self._op_buffer = None
+        self._materialized_data = None
+        self._meta_data = None
 
-                # apply cached sequence
-                self._cached_data = self._apply_cache_buffer(func(*args, **kwargs))
-            else:
-                # apply cached sequence only
-                self._cached_data = self._apply_cache_buffer()
-        return self._cached_data
-
-    def _apply_cache_buffer(self, target=None) -> torch.Tensor:
+    @staticmethod
+    def _replace_with_materialized(x):
+        if isinstance(x, LazyTensor):
+            return x._materialize_data()
+        return x
+
+    def _materialize_data(self) -> torch.Tensor:
+        # self._materialized_data should be generated after the first call of this function
+        if self._materialized_data is None:
+            # apply factory method
+            func, args, kwargs = self._factory_method
+
+            # apply cached sequence
+            self._pre_op_fn()
+
+            try:
+                init_val = func(*tree_map(self._replace_with_materialized, args),
+                                **tree_map(self._replace_with_materialized, kwargs))
+            except TypeError as e:
+                print(f'init fn: {func.__name__}')
+                raise e
+
+            self._materialized_data = self._rerun_ops(init_val)
+        return self._materialized_data
+
+    def _rerun_ops(self, target=None) -> torch.Tensor:
```

Comment:
Add docstring to explain the definition of `rerun`.

Reply:
done